lib/engine/part/path.rb

```diff
@@ -51,10 +50,36 @@ def <=>(other)
       end
 
       def <=(other)
+        other_ends = other.ends
+        ends.all? { |t| other_ends.any? { |o| t <= o } }
+      end
+
+      def match?(other)
         (@a <= other.a && @b <= other.b) ||
           (@a <= other.b && @b <= other.a)
       end
 
+      def ends
+        @ends ||= calculate_ends
+      end
+
+      def calculate_ends
```

Comment:
this could be

[@a, @b].flat_map do |j|
  next [j] unless j.junction?

  j.paths.map do |p|
    next if p == self
    [p.a, p.b].select(&:junction?)
  end

Reply:
Ok